mysys/thr_lock.c

```diff
@@ -546,9 +546,11 @@ wait_for_lock(struct st_lock_list *wait, THR_LOCK_DATA *data,
   statistic_increment(locks_waited, &THR_LOCK_lock);
 
   /* Set up control struct to allow others to abort locks */
+  mysql_mutex_lock(&thread_var->mutex);
   thread_var->current_mutex= &data->lock->mutex;
   thread_var->current_cond=  cond;
   data->cond= cond;
+  mysql_mutex_unlock(&thread_var->mutex);
```

Comment:
This is not great, because we have to lock mutexes (`&thread_var->mutex` and `thread_var->current_mutex`) in opposite order when killing connection. You replaced it with `trylock()` to avoid deadlock. But then we may miss to broadcast "awake" to victim thread.

I have no good solution on my mind atm.